processor/src/host/advice/mod.rs

```diff
@@ -0,0 +1,727 @@
+    ///
+    /// For more information, check [MerkleStore::get_leaf_depth].
+    ///
+    /// # Errors
+    /// Will return an error if:
+    /// - The provided `tree_depth` doesn't fit `u8`.
+    /// - The conditions of [MerkleStore::get_leaf_depth] aren't met.
+    fn get_leaf_depth(
+        &self,
+        root: Word,
+        tree_depth: &Felt,
+        index: &Felt,
+    ) -> Result<u8, ExecutionError>;
+
+    /// Returns node value and index of a leaf node in the subtree of the specified root, if and
+    /// only if this is the only leaf in the entire subtree. Otherwise, None is returned.
+    ///
+    /// The root itself is assumed to be located at the specified index in a tree with the provided
+    /// depth.
+    ///
+    /// # Errors
+    /// Returns an error if a three for the specified root does not exist in the advice provider.
+    fn find_lone_leaf(
+        &self,
+        root: Word,
+        root_index: NodeIndex,
+        tree_depth: u8,
+    ) -> Result<Option<(NodeIndex, Word)>, ExecutionError>;
+
+    /// Updates a node at the specified depth and index in a Merkle tree with the specified root;
+    /// returns the Merkle path from the updated node to the new root, together with the new root.
+    ///
+    /// The tree is cloned prior to the update. Thus, the advice provider retains the original and
+    /// the updated tree.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - A Merkle tree for the specified root cannot be found in this advice provider.
+    /// - The specified depth is either zero or greater than the depth of the Merkle tree
+    ///   identified by the specified root.
+    /// - Path to the leaf at the specified index in the specified Merkle tree is not known to this
+    ///   advice provider.
+    fn update_merkle_node(
+        &mut self,
+        root: Word,
+        depth: &Felt,
+        index: &Felt,
+        value: Word,
+    ) -> Result<(MerklePath, Word), ExecutionError>;
+
+    /// Creates a new Merkle tree in the advice provider by combining Merkle trees with the
+    /// specified roots. The root of the new tree is defined as `hash(left_root, right_root)`.
+    ///
+    /// After the operation, both the original trees and the new tree remains in the advice
+    /// provider (i.e., the input trees are not removed).
+    ///
+    /// # Errors
+    /// Returns an error if a Merkle tree for either of the specified roots cannot be found in this
+    /// advice provider.
+    fn merge_roots(&mut self, lhs: Word, rhs: Word) -> Result<Word, ExecutionError>;
+
+    /// Returns a subset of this Merkle store such that the returned Merkle store contains all
+    /// nodes which are descendants of the specified roots.
+    ///
+    /// The roots for which no descendants exist in this Merkle store are ignored.
+    fn get_store_subset<I, R>(&self, roots: I) -> MerkleStore
+    where
+        I: Iterator<Item = R>,
+        R: Borrow<RpoDigest>;
```

Comment:
Upon review I think it may be possible to do so using the `get_tree_node(..)` method of the advice provider interface however we will be duplicating a lot of logic that is already within the `MerkleStore` which is nicely encapsulated.  It's a better fit there in my opinion. I would suggest we leave this change out of scope for this PR and open an issue to consider this at some point in the future.